### PR TITLE
Add option for setting "preferred" backend

### PR DESCRIPTION
**Pull request by nelson-liu**
This PR lets you specify a preferred backend in the json, and a warning will be displayed if the experiment is run with a different backend (but it'll still run).

this is motivated by my recent experience with the attention sum reader --- the theano model trains quite well and gets around 0.55 keras val accuracy (comparable to published result), but the tensorflow backend doesn't seem to learn anything and just stagnates at 0.38 accuracy. Thus, as the writer of the experiment i'd like to strongly using theano to run it, and i would set this parameter.

### CODE REVIEW

File: src/main/python/deep_qa/training/trainer.py [hunk 1]
```diff
@@ -42,6 +42,13 @@ def __init__(self, params: Dict[str, Any]):
             parent_directory = os.path.dirname(self.model_prefix)
             os.makedirs(parent_directory, exist_ok=True)
 
+        # Preferred backend to use for training. If a different backend is detected, we still
+        # train but we also warn the user.
+        self.preferred_backend = params.pop('preferred_backend', None)
+        if self.preferred_backend and self.preferred_backend != K.backend():
+            logger.warning("Preferred backend is %s, but current backend is %s.",
```
Comment (matt-gardner) [inline src/main/python/deep_qa/training/trainer.py]: Is it worth waiting here for user input?  If not I would probably take several lines to display this warning, making it really bold and obvious.

Also, should you update your example experiment with the ASReader to include this parameter?

**Comment by nelson-liu** [comment]
@matt-gardner done, here's an image of what the new warning looks like (thanks, ssh!)

screen shot 2017-01-26 at 5 33 26 pm

File: src/main/python/deep_qa/training/trainer.py [hunk 2]
```diff
@@ -42,6 +42,27 @@ def __init__(self, params: Dict[str, Any]):
             parent_directory = os.path.dirname(self.model_prefix)
             os.makedirs(parent_directory, exist_ok=True)
 
+        # Preferred backend to use for training. If a different backend is detected, we still
+        # train but we also warn the user.
+        self.preferred_backend = params.pop('preferred_backend', None)
+        if self.preferred_backend and self.preferred_backend.lower() != K.backend():
+            warning_info = ("@ Preferred backend is %s, but "
```
Comment (matt-gardner) [inline src/main/python/deep_qa/training/trainer.py]: One last nitpick - this is a lot of logic to be sitting in `__init__`.  Can you move it to a function like `_print_backend_warning()`?
